File: scripts/add_marker.py
#!/usr/bin/env python3
"""
Marker Addition Script for Daemon X Machina: Titanic Scion Interactive Map

This script adds markers to GeoJSON files for the interactive map.
Usage: python scripts/add_marker.py {map_id} {category} "{name}" {x} {y}
"""

import argparse
import json
import sys
from pathlib import Path


# Valid categories based on existing project structure
# NOTE: This list must be synchronized with the 'colors' object in src/icons.js
# Any changes here should be reflected in the frontend icon system
VALID_CATEGORIES = ['bgm', 'card', 'chest', 'decal', 'enemy', 'log']

# Default paths
DEFAULT_MARKERS_DIR = 'assets/data/markers'


def parse_arguments():
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(
        description='Add markers to GeoJSON files for the interactive map',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  python scripts/add_marker.py forest card "Rare Card" 800 600
  python scripts/add_marker.py desert chest "Hidden Treasure" 400 300 --dry-run
  python scripts/add_marker.py --categories
        """
    )
    
    # Optional arguments that can be used standalone
    parser.add_argument('--dry-run', action='store_true', 
                       help='Show what would be added without modifying files')
    parser.add_argument('--categories', action='store_true',
                       help='List valid categories and exit')
    
    # Required arguments (only when not using --categories)
    parser.add_argument('map_id', nargs='?', help='Map identifier. Available maps: forest, desert, mountains')
    parser.add_argument('category', nargs='?', choices=VALID_CATEGORIES, 
                       help=f'Marker category. Valid options: {", ".join(VALID_CATEGORIES)}')
    parser.add_argument('name', nargs='?', help='Marker name/title')
    parser.add_argument('x', nargs='?', type=int, help='X coordinate (horizontal)')
    parser.add_argument('y', nargs='?', type=int, help='Y coordinate (vertical)')
    
    args = parser.parse_args()
    
    # Validate arguments based on usage
    if args.categories:
        return args
    
    # Check that all required arguments are provided when not using --categories
    missing_args = []
    if args.map_id is None:
        missing_args.append('map_id')
    if args.category is None:
        missing_args.append('category')
    if args.name is None:
        missing_args.append('name')
    if args.x is None:
        missing_args.append('x')
    if args.y is None:
        missing_args.append('y')
    
    if missing_args:
        parser.error(f"the following arguments are required: {', '.join(missing_args)}")
    
    return args


def show_categories():
    """Display valid categories and exit."""
    print("Valid marker categories:")
    for category in VALID_CATEGORIES:
        print(f"  - {category}")
    sys.exit(0)


def get_file_path(map_id):
    """Get the file path for the marker data."""
    return Path(DEFAULT_MARKERS_DIR) / f"{map_id}.geojson"


def validate_inputs(args):
    """Validate input arguments."""
    # Validate coordinates are non-negative
    if args.x < 0 or args.y < 0:
        print(f"Error: Coordinates must be non-negative. Got x={args.x}, y={args.y}", file=sys.stderr)
        sys.exit(1)
    
    # Validate marker name is not empty
    if not args.name.strip():
        print("Error: Marker name cannot be empty", file=sys.stderr)
        sys.exit(1)


def load_geojson(file_path):
    """Load existing GeoJSON file or create new FeatureCollection structure."""
    if file_path.exists():
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                data = json.load(f)
            
            # Validate GeoJSON structure
            if not isinstance(data, dict) or data.get('type') != 'FeatureCollection':
                raise ValueError("File is not a valid GeoJSON FeatureCollection")
            
            if 'features' not in data or not isinstance(data['features'], list):
                raise ValueError("Invalid GeoJSON: 'features' must be a list")
            
            return data
        
        except json.JSONDecodeError as e:
            raise ValueError(f"Invalid JSON format: {e}")
        except Exception as e:
            raise ValueError(f"Error reading file: {e}")
    
    else:
        # Create new FeatureCollection structure
        return {
            "type": "FeatureCollection",
            "features": []
        }


def save_geojson(file_path, data):
    """Save GeoJSON data to file with pretty formatting."""
    # Ensure parent directory exists
    file_path.parent.mkdir(parents=True, exist_ok=True)
    
    try:
        with open(file_path, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
            f.write('\n')  # Add trailing newline
        
        return True
    
    except Exception as e:
        raise ValueError(f"Error writing file: {e}")


def extract_existing_data(geojson_data):
    """Extract existing IDs, coordinates, and names from GeoJSON."""
    existing_ids = set()
    existing_coords = set()
    existing_names = set()
    
    for feature in geojson_data.get('features', []):
        # Extract ID
        if 'properties' in feature and 'id' in feature['properties']:
            existing_ids.add(feature['properties']['id'])
        
        # Extract coordinates
        if 'geometry' in feature and 'coordinates' in feature['geometry']:
            coords = feature['geometry']['coordinates']
            if len(coords) >= 2:
                existing_coords.add((int(coords[0]), int(coords[1])))
        
        # Extract name
        if 'properties' in feature and 'name' in feature['properties']:
            existing_names.add(feature['properties']['name'])
    
    return existing_ids, existing_coords, existing_names


def generate_next_id(map_id, existing_ids):
    """Generate next available ID in format {map_id}-{3-digit number}."""
    prefix = f"{map_id}-"
    max_number = 0
    
    # Find the highest existing number for this map
    for existing_id in existing_ids:
        if existing_id.startswith(prefix):
            suffix = existing_id[len(prefix):]
            if suffix.isdigit():
                max_number = max(max_number, int(suffix))
    
    # Generate next ID with 3-digit format to match existing convention
    next_number = max_number + 1
    return f"{prefix}{next_number:03d}"


def check_duplicates(marker_id, x, y, name, existing_ids, existing_coords, existing_names):
    """Check for duplicate IDs, coordinates, and names."""
    warnings = []
    
    # Check ID duplication
    if marker_id in existing_ids:
        warnings.append(f"ID '{marker_id}' already exists")
    
    # Check coordinate duplication
    coord_tuple = (x, y)
    if coord_tuple in existing_coords:
        warnings.append(f"Coordinates ({x}, {y}) already exist")
    
    # Check name duplication
    if name in existing_names:
        warnings.append(f"Name '{name}' already exists")
    
    if warnings:
        print("Error: Duplicate data detected:", file=sys.stderr)
        for warning in warnings:
            print(f"  - {warning}", file=sys.stderr)
        return False
    
    return True


def create_marker_feature(marker_id, name, category, x, y):
    """Create a GeoJSON Feature for the marker."""
    return {
        "type": "Feature",
        "geometry": {
            "type": "Point",
            "coordinates": [x, y]
        },
        "properties": {
            "id": marker_id,
            "name": name,
            "category": category
        }
    }


def add_marker_to_geojson(geojson_data, marker_feature):
    """Add marker feature to GeoJSON FeatureCollection."""
    if 'features' not in geojson_data:
        geojson_data['features'] = []
    
    geojson_data['features'].append(marker_feature)
    
    # Sort features by ID for consistent ordering
    geojson_data['features'].sort(key=lambda f: f.get('properties', {}).get('id', ''))
    
    return geojson_data


def display_marker_preview(marker_feature, file_path):
    """Display preview of the marker that would be added."""
    props = marker_feature['properties']
    coords = marker_feature['geometry']['coordinates']
    
    print("\nMarker to be added:")
    print(f"  ID: {props['id']}")
    print(f"  Name: {props['name']}")
    print(f"  Category: {props['category']}")
    print(f"  Coordinates: ({coords[0]}, {coords[1]})")
    print(f"  Target file: {file_path}")
    
    # Display JSON representation
    print("\nJSON representation:")
    print(json.dumps(marker_feature, indent=2, ensure_ascii=False))


def main():
    """Main function."""
    args = parse_arguments()
    
    # Handle --categories option
    if args.categories:
        show_categories()
    
    # Validate inputs
    validate_inputs(args)
    
    try:
        # Get file path
        file_path = get_file_path(args.map_id)
        
        print(f"Target file: {file_path}")
        print(f"Map ID: {args.map_id}")
        print(f"Category: {args.category}")
        print(f"Name: {args.name}")
        print(f"Coordinates: ({args.x}, {args.y})")
        
        # Load existing GeoJSON data
        print(f"\nLoading GeoJSON data from {file_path}...")
        geojson_data = load_geojson(file_path)
        
        # Extract existing data for duplicate checking
        existing_ids, existing_coords, existing_names = extract_existing_data(geojson_data)
        print(f"Found {len(existing_ids)} existing markers")
        
        # Generate marker ID
        marker_id = generate_next_id(args.map_id, existing_ids)
        print(f"Generated ID: {marker_id}")
        
        # Check for duplicates
        if not check_duplicates(marker_id, args.x, args.y, args.name, 
                               existing_ids, existing_coords, existing_names):
            sys.exit(1)
        
        # Create marker feature
        marker_feature = create_marker_feature(marker_id, args.name, args.category, args.x, args.y)
        
        # Handle dry run
        if args.dry_run:
            print("\n[DRY RUN] No files will be modified")
            display_marker_preview(marker_feature, file_path)
            return
        
        # Add marker to GeoJSON
        updated_geojson = add_marker_to_geojson(geojson_data, marker_feature)
        
        # Save updated GeoJSON
        print(f"\nSaving updated GeoJSON to {file_path}...")
        save_geojson(file_path, updated_geojson)
        
        print(f"✓ Marker '{args.name}' successfully added with ID '{marker_id}'")
        print(f"✓ Total markers in file: {len(updated_geojson['features'])}")
        
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == '__main__':
    main()

File: scripts/add_marker.js
import fs from 'node:fs';
import path from 'node:path';
import process from 'node:process';
import { pathToFileURL } from 'node:url';
import categoryColors from "../src/data/category-colors.json" assert { type: 'json' };

const colors = /** @type {Record<string, string>} */ (categoryColors);

/**
 * @typedef {Object} FeatureGeometry
 * @property {string} [type]
 * @property {(number[] | [number, number])} [coordinates]
 */

/**
 * @typedef {Object} FeatureLike
 * @property {string} [type]
 * @property {FeatureGeometry} [geometry]
 * @property {Record<string, unknown>} [properties]
 */

/**
 * @typedef {Object} FeatureCollectionLike
 * @property {"FeatureCollection"} type
 * @property {FeatureLike[]} features
 */

/**
 * @typedef {Object} MarkerArgumentsBase
 * @property {boolean} dryRun
 */

/**
 * @typedef {MarkerArgumentsBase & {
 *   categoriesFlag: false;
 *   mapId: string;
 *   x: number;
 *   y: number;
 *   category: string;
 *   name: string;
 *   description: string;
 * }} MarkerArguments
 */

/**
 * @typedef {MarkerArgumentsBase & {
 *   categoriesFlag: true;
 *   mapId: undefined;
 *   x: undefined;
 *   y: undefined;
 *   category: undefined;
 *   name: undefined;
 *   description: "";
 * }} CategoryListArguments
 */

/**
 * @typedef {MarkerArguments | CategoryListArguments} ParsedArguments
 */

/**
 * @typedef {Object} MarkerInput
 * @property {boolean} [dryRun]
 * @property {string} [mapId]
 * @property {unknown} x
 * @property {unknown} y
 * @property {unknown} category
 * @property {unknown} name
 * @property {unknown} [description]
 */

/**
 * @typedef {Object} ValidatedMarkerInput
 * @property {boolean} [dryRun]
 * @property {string} [mapId]
 * @property {number} x
 * @property {number} y
 * @property {string} category
 * @property {string} name
 * @property {string} description
 */

/**
 * @typedef {Object} MarkerFeature
 * @property {"Feature"} type
 * @property {{ type: "Point"; coordinates: [number, number] }} geometry
 * @property {{ id: string; name: string; category: string; description?: string }} properties
 */

export const VALID_CATEGORIES = Object.freeze(Object.keys(colors));

export const DEFAULT_MARKERS_DIR = 'public/assets/data/markers';

const COORD_SEPARATOR = ':';

/**
 * @param {number} x
 * @param {number} y
 * @returns {string}
 */
const toCoordKey = (x, y) => `${Number(x)}${COORD_SEPARATOR}${Number(y)}`;

/**
 * @param {string[]} [argv]
 * @returns {ParsedArguments}
 */
export function parseArguments(argv = process.argv.slice(2)) {
  const options = { dryRun: false, categoriesFlag: false };
  const positional = [];

  for (const token of argv) {
    if (token === '--dry-run') {
      options.dryRun = true;
      continue;
    }

    if (token === '--categories') {
      options.categoriesFlag = true;
      continue;
    }

    positional.push(token);
  }

  if (options.categoriesFlag) {
    return {
      dryRun: options.dryRun,
      categoriesFlag: true,
      mapId: undefined,
      x: undefined,
      y: undefined,
      category: undefined,
      name: undefined,
      description: '',
    };
  }

  if (positional.length < 5) {
    const missing = ['map_id', 'x', 'y', 'category', 'name'].slice(positional.length);
    throw new Error(`Missing required arguments: ${missing.join(', ')}`);
  }

  const [mapId, xRaw, yRaw, category, name, ...descriptionParts] = positional;

  const x = Number.parseInt(xRaw, 10);
  const y = Number.parseInt(yRaw, 10);

  if (!Number.isFinite(x) || Number.isNaN(x)) {
    throw new Error(`Invalid x coordinate: ${xRaw}`);
  }

  if (!Number.isFinite(y) || Number.isNaN(y)) {
    throw new Error(`Invalid y coordinate: ${yRaw}`);
  }

  const description = descriptionParts.length > 0 ? descriptionParts.join(' ') : '';

  return {
    dryRun: options.dryRun,
    categoriesFlag: false,
    mapId,
    x,
    y,
    category,
    name,
    description,
  };
}

/**
 * @param {readonly string[]} [categories]
 * @returns {void}
 */
export function showCategories(categories = VALID_CATEGORIES) {
  console.log('Valid marker categories:');
  for (const category of categories) {
    console.log(`  - ${category}`);
  }
}

/**
 * @param {string} mapId
 * @param {string} [markersDir]
 * @returns {string}
 */
export function getFilePath(mapId, markersDir = DEFAULT_MARKERS_DIR) {
  if (!mapId) {
    throw new Error('Map ID is required');
  }

  return path.resolve(markersDir, `${mapId}.geojson`);
}

/**
 * @param {MarkerInput} args
 * @returns {ValidatedMarkerInput}
 */
export function validateInputs(args) {
  if (typeof args.x !== 'number' || typeof args.y !== 'number') {
    throw new Error('Coordinates must be numbers');
  }

  if (args.x < 0 || args.y < 0) {
    throw new Error(`Coordinates must be non-negative. Got x=${args.x}, y=${args.y}`);
  }

  if (typeof args.category !== 'string') {
    throw new Error(`Invalid category: ${String(args.category)}`);
  }

  if (!VALID_CATEGORIES.includes(args.category)) {
    throw new Error(`Invalid category: ${args.category}`);
  }

  const trimmedName = typeof args.name === 'string' ? args.name.trim() : '';
  if (!trimmedName) {
    throw new Error('Marker name cannot be empty');
  }

  const trimmedDescription = typeof args.description === 'string' ? args.description.trim() : '';

  const mapId = typeof args.mapId === 'string' ? args.mapId : undefined;

  return /** @type {ValidatedMarkerInput} */ ({
    dryRun: Boolean(args.dryRun),
    mapId,
    x: args.x,
    y: args.y,
    category: args.category,
    name: trimmedName,
    description: trimmedDescription,
  });
}

/**
 * @param {import('node:fs').PathLike} filePath
 * @returns {FeatureCollectionLike}
 */
export function loadGeojson(filePath) {
  if (fs.existsSync(filePath)) {
    try {
      const raw = fs.readFileSync(filePath, 'utf-8');
      const data = JSON.parse(raw);

      if (!data || data.type !== 'FeatureCollection') {
        throw new Error('File is not a valid GeoJSON FeatureCollection');
      }

      if (!Array.isArray(data.features)) {
        throw new Error("Invalid GeoJSON: 'features' must be an array");
      }

      return data;
    } catch (error) {
      if (error instanceof SyntaxError) {
        throw new Error(`Invalid JSON format: ${error.message}`);
      }

      const message = error instanceof Error ? error.message : String(error);
      throw new Error(`Error reading file: ${message}`);
    }
  }

  return /** @type {FeatureCollectionLike} */ ({
    type: 'FeatureCollection',
    features: [],
  });
}

/**
 * @param {import('node:fs').PathLike} filePath
 * @param {FeatureCollectionLike} data
 * @returns {true}
 */
export function saveGeojson(filePath, data) {
  if (typeof filePath !== 'string') {
    throw new TypeError('filePath must be a string path');
  }

  fs.mkdirSync(path.dirname(filePath), { recursive: true });
  try {
    fs.writeFileSync(filePath, `${JSON.stringify(data, null, 2)}\n`, 'utf-8');
    return true;
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    throw new Error(`Error writing file: ${message}`);
  }
}

/**
 * @param {{ features?: FeatureLike[] }} geojsonData
 * @returns {{ existingIds: Set<string>; existingCoords: Set<string> }}
 */
export function extractExistingData(geojsonData) {
  const existingIds = new Set();
  const existingCoords = new Set();

  for (const feature of geojsonData.features ?? []) {
    const properties = feature?.properties ?? {};
    const geometry = feature?.geometry ?? {};

    if (typeof properties.id === 'string') {
      existingIds.add(properties.id);
    }

    if (Array.isArray(geometry.coordinates) && geometry.coordinates.length >= 2) {
      const [x, y] = geometry.coordinates;
      existingCoords.add(toCoordKey(x, y));
    }

  }

  return { existingIds, existingCoords };
}

/**
 * @param {string} mapId
 * @param {Iterable<string>} existingIds
 * @returns {string}
 */
export function generateNextId(mapId, existingIds) {
  const prefix = `${mapId}-`;
  let maxNumber = 0;

  for (const existingId of existingIds) {
    if (existingId.startsWith(prefix)) {
      const suffix = existingId.slice(prefix.length);
      if (/^\d+$/.test(suffix)) {
        maxNumber = Math.max(maxNumber, Number.parseInt(suffix, 10));
      }
    }
  }

  const nextNumber = maxNumber + 1;
  return `${prefix}${nextNumber.toString().padStart(3, '0')}`;
}

/**
 * @param {string} markerId
 * @param {number} x
 * @param {number} y
 * @param {Set<string>} existingIds
 * @param {Set<string>} existingCoords
 * @returns {string[]}
 */
export function checkDuplicates(markerId, x, y, existingIds, existingCoords) {
  const warnings = [];

  if (existingIds.has(markerId)) {
    warnings.push(`ID '${markerId}' already exists`);
  }

  if (existingCoords.has(toCoordKey(x, y))) {
    warnings.push(`Coordinates (${x}, ${y}) already exist`);
  }

  return warnings;
}

/**
 * @param {string} markerId
 * @param {string} name
 * @param {string} category
 * @param {number} x
 * @param {number} y
 * @param {string} [description]
 * @returns {MarkerFeature}
 */
export function createMarkerFeature(markerId, name, category, x, y, description = '') {
  /** @type {MarkerFeature} */
  const feature = {
    type: 'Feature',
    geometry: {
      type: 'Point',
      coordinates: [x, y],
    },
    properties: {
      id: markerId,
      name,
      category,
    },
  };

  if (description) {
    feature.properties.description = description;
  }

  return feature;
}

/**
 * @template T
 * @param {T & { features?: FeatureLike[] }} geojsonData
 * @param {FeatureLike} markerFeature
 * @returns {T & { features: FeatureLike[] }}
 */
export function addMarkerToGeojson(geojsonData, markerFeature) {
  if (!Array.isArray(geojsonData.features)) {
    geojsonData.features = [];
  }

  geojsonData.features.push(markerFeature);
  geojsonData.features.sort((a, b) => {
    const idA = typeof a?.properties?.id === 'string' ? a.properties.id : '';
    const idB = typeof b?.properties?.id === 'string' ? b.properties.id : '';
    return idA.localeCompare(idB);
  });

  return /** @type {T & { features: FeatureLike[] }} */ (geojsonData);
}

/**
 * @param {MarkerFeature} markerFeature
 * @param {string} filePath
 * @returns {void}
 */
export function displayMarkerPreview(markerFeature, filePath) {
  const { properties, geometry } = markerFeature;
  console.log('\nMarker to be added:');
  console.log(`  ID: ${properties.id}`);
  console.log(`  Name: ${properties.name}`);
  console.log(`  Category: ${properties.category}`);
  console.log(`  Coordinates: (${geometry.coordinates[0]}, ${geometry.coordinates[1]})`);
  if (properties.description) {
    console.log(`  Description: ${properties.description}`);
  }
  console.log(`  Target file: ${filePath}`);
  console.log('\nJSON representation:');
  console.log(JSON.stringify(markerFeature, null, 2));
}

/**
 * @param {string[]} [argv]
 * @param {string} [markersDir]
 * @returns {void}
 */
export function main(argv = process.argv.slice(2), markersDir = DEFAULT_MARKERS_DIR) {
  const args = parseArguments(argv);

  if (args.categoriesFlag) {
    showCategories();
    return;
  }

  const validated = validateInputs(args);
  if (!validated.mapId) {
    throw new Error('Map ID is required');
  }

  const filePath = getFilePath(validated.mapId, markersDir);

  console.log(`Target file: ${filePath}`);
  console.log(`Map ID: ${validated.mapId}`);
  console.log(`Coordinates: (${validated.x}, ${validated.y})`);
  console.log(`Category: ${validated.category}`);
  console.log(`Name: ${validated.name}`);
  if (validated.description) {
    console.log(`Description: ${validated.description}`);
  }

  console.log(`\nLoading GeoJSON data from ${filePath}...`);
  const geojsonData = loadGeojson(filePath);
  const { existingIds, existingCoords } = extractExistingData(geojsonData);
  console.log(`Found ${existingIds.size} existing markers`);

  const markerId = generateNextId(validated.mapId, existingIds);
  console.log(`Generated ID: ${markerId}`);

  const warnings = checkDuplicates(
    markerId,
    validated.x,
    validated.y,
    existingIds,
    existingCoords,
  );

  if (warnings.length > 0) {
    console.error('Error: Duplicate data detected:');
    for (const warning of warnings) {
      console.error(`  - ${warning}`);
    }
    process.exitCode = 1;
    return;
  }

  const markerFeature = createMarkerFeature(
    markerId,
    validated.name,
    validated.category,
    validated.x,
    validated.y,
    validated.description,
  );

  if (validated.dryRun) {
    console.log('\n[DRY RUN] No files will be modified');
    displayMarkerPreview(markerFeature, filePath);
    return;
  }

  const updatedGeojson = addMarkerToGeojson(geojsonData, markerFeature);

  console.log(`\nSaving updated GeoJSON to ${filePath}...`);
  saveGeojson(filePath, updatedGeojson);

  console.log(`✓ Marker '${validated.name}' successfully added with ID '${markerId}'`);
  console.log(`✓ Total markers in file: ${updatedGeojson.features.length}`);
}

/**
 * @param {string | null | undefined} [entry]
 * @param {string} [moduleUrl]
 * @returns {boolean}
 */
export function isExecutedDirectly(entry = process.argv[1], moduleUrl = import.meta.url) {
  if (!entry) {
    return false;
  }
  try {
    return pathToFileURL(entry).href === moduleUrl;
  } catch {
    return false;
  }
}

/* c8 ignore start */
const runFromCli = isExecutedDirectly();

if (runFromCli) {
  try {
    main();
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    console.error(`Error: ${message}`);
    process.exit(1);
  }
}
/* c8 ignore end */
